mne/preprocessing/maxwell.py

```diff
@@ -0,0 +1,255 @@
+from scipy.misc import factorial
+from ..forward._lead_dots import (_get_legen_table, _get_legen_lut_fast,
+                                  _get_legen_lut_accurate)
+from ..fixes import partial
+from ..transforms import _cartesian_to_sphere as cart_to_sph
+from ..transforms import _sphere_to_cartesian as sph_to_cart
+from scipy.misc import factorial as fact
+
+
+def maxwell_filter(Raw, origin, L_in, L_out):
+    """
+    Filter data using SSS basis functions.
+
+    Parameters
+    ----------
+    Raw : mne.Raw instance
+        Data to be filtered.
+    origin : array of shape [3,]
+        Origin of head.
+    L_in : int
+        Order of in-component spherical expansion
+    L_out : int
+        Order of out-component spherical expansion
+
+    Returns
+    -------
+    sss_Raw : Filtered data
+    """
+
+    #TODO: Figure out all parameters required
+    #TODO: Error checks on input parameters
+
+    #TODO: Compute spherical harmonics
+    #TODO: Project data into spherical harmonics space
+    #TODO: Reconstruct and return Raw file object
+
+
+def sph_harmonic(l, m, az, pol):
+    """
+    Compute the spherical harmonic function at point in spherical coordinates.
+    When using, pay close attention to inputs. Spherical harmonic notation for
+    order/degree, and theta/phi are both reversed in original SSS work compared
+    to many other sources.
+
+    Parameters
+    ----------
+    l : int
+        Degree of spherical harmonic
+    m : int
+        Order of spherical harmonic
+    az : float
+        Azimuthal (longitudinal) spherical coordinate [0, 2*pi]. 0 is aligned
+        with x-axis.
+    pol : float
+        Polar (or colatitudinal) spherical coordinate [0, pi]. 0 is aligned
+        with z-axis.
+
+    Returns
+    -------
+    y : complex float
+        The spherical harmonic value at the specified polar and azimuth angles
+    """
+    assert np.abs(m) <= l, 'Absolute value of expansion coefficient must be <= l'
+
+    # Get function for Legendre derivatives
+    #lut, n_fact = _get_legen_table('meg', False, 100)
+    #lut_deriv_fun = partial(_get_legen_lut_accurate, lut=lut)
+
+    #TODO: Decide on notation for spherical coords
+    #TODO: Should factorial function use floating or long precision?
+
+    #Ensure that polar and azimuth angles are
+    polar = np.array(pol)
+    azimuth = np.array(az)
+
+    # Real valued spherical expansion as given by Wikso, (11)
+    base = np.sqrt(((2 * l + 1) / (4 * np.pi) *
+                    factorial(l - m) / factorial(l + m)) *
+                   lpmv(m, l, np.cos(polar)))
+    if m < 0:
+        return base * np.sin(m * azimuth)
+    else:
+        return base * np.cos(m * azimuth)
+
+    #TODO: Check speed of taking real part of scipy's sph harmonic function
+    # Note reversal in notation order between scipy and original SSS papers
+    # Degree/order and theta/phi reversed
+    #return np.real(sph_harm(m, l, azimuth, polar))
+
+
+def get_num_harmonics(L_in, L_out):
```

Comment:
eventually most of these functions should be privatized by having a leading underscore